From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: kernel: diagnose 10 does not release memory above 2GB
References: bnc#691408,LTC#71847
Patch-mainline: Yes

Symptom:     The working set size of active pages as reported by z/VM is
             not consistent with the size of the CMM balloon.
Problem:     The diag10 function does nothing if the address of the page
             to release is higher than 2047MB because the diagnose X'10'
             interface of z/VM prior to release 5.2 had been restricted
             to the 31bit addressing mode and to addresses below 2047MB.
Solution:    Remove addressing mode switch to 31 bit and allow addresses
             above 2047MB. Add two expection table entries and continue
             after the diagnose instruction in the case it causes program
             checks.

Acked-by: John Jolly <jjolly@suse.de>

---
 arch/s390/kernel/diag.c |   17 +++++------------
 1 file changed, 5 insertions(+), 12 deletions(-)

--- a/arch/s390/kernel/diag.c
+++ b/arch/s390/kernel/diag.c
@@ -13,19 +13,12 @@
  */
 void diag10(unsigned long addr)
 {
-	if (addr >= 0x7ff00000)
-		return;
 	asm volatile(
-#ifdef CONFIG_64BIT
-		"	sam31\n"
-		"	diag	%0,%0,0x10\n"
-		"0:	sam64\n"
-#else
-		"	diag	%0,%0,0x10\n"
-		"0:\n"
-#endif
-		EX_TABLE(0b, 0b)
-		: : "a" (addr));
+		"0:	diag	%0,%0,0x10\n"
+		"1:\n"
+		EX_TABLE(0b, 1b)
+		EX_TABLE(1b, 1b)
+  		: : "a" (addr));
 }
 EXPORT_SYMBOL(diag10);
 
